### PR TITLE
Detail best practices for working with CocoaPods on larger projects

### PR DESCRIPTION
**Pull request by alexpersian**
Addresses issue #141 

Added a file with information on best practices. This seemed like something that was missing from Sherpa. I'd like to get team input on the stuff detailed here, especially from people who have dealt with more complex pods setups or challenges around CocoaPods.

### CODE REVIEW

File: ios/cocoapods_best_practices.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    pod 'Analytics'
+    pod 'Crashlytics'
+    pod 'OHHTTPStubs/Swift'
+    pod 'Firebase/Core'
+    pod 'Alamofire'
+    pod 'Fabric', '~> 1.7'
+    pod 'IGListKit', '~> 3.1'
+end
+```
+This Podfile is very quick to set up, and very easy to use. That ease comes with a tradeoff though where you aren't guaranteeing the version that will be installed when `pod install` is run. Sure, for Fabric and IGListKit you are giving a version, but they will bring in any version up to `2.0` and `4.0` respectively. Granted the maintainers should be adhering to semver style and shouldn't be introducing any major breaking changes in those versions, but in the event of that happening your project could be broken unexpectedly.
```
Comment (Benuuu) [inline ios/cocoapods_best_practices.md]: expand "semver"

Comment (Benuuu) [inline ios/cocoapods_best_practices.md]: > That ease comes with a tradeoff though where you aren't guaranteeing the version that will be installed when `pod install` is run.

The versioning is well defined, and there may be a case where you want to install any version up to and including that.  Your pod will not automatically update the version on `pod install` if you have a `Podfile.lock` in the repository.  It only updates on `pod upgrade`

https://guides.cocoapods.org/using/pod-install-vs-update.html

File: ios/cocoapods_best_practices.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+- Updating the workspace, such as a new scheme or configuration
+- Something broke and the Pods project needs to be refreshed
+
+In those cases you'll want to run `pod install` again to generate a new lock file, and sync with your team about its changes.
+
+## Checking in your Pods
+
+Imagine this scenario. You're bringing a new developer onto the project or you are versioning a new workstation and need to install all of the project pods from scratch. The project has a well defined podfile so you aren't worried about version mismatches or breaking changes from new releases. When you run `pod install` though CocoaPods throws an error saying that a particular pod at the core of your project isn't found. It was removed from the database and is no longer available. Not likely right? You'd think, but that's [exactly what happened in the Node world](https://blog.npmjs.org/post/141577284765/kik-left-pad-and-npm) not long ago. **This is why it's so important to include your pods in the git repo when checking it in**. If the worst case scenario happens the project won't be dead in the water. Sure it takes longer to clone the repo at first, but the extra time is well worth the reassurance.
```
Comment (Benuuu) [inline ios/cocoapods_best_practices.md]: Open with "Prefer to check in your `Pods/` directory" as the rest is an explanation of why/what.

Or change the header to be more explicit "Prefer to commit your pods".

File: ios/cocoapods_best_practices.md (hunk 1 above)
Comment (alexpersian) [reply ios/cocoapods_best_practices.md]: Will do, I'll add a link to the semver documentation.

Comment (alexpersian) [reply ios/cocoapods_best_practices.md]: Oh interesting. I didn't know about the differences between those two. This only adds an extra guarantee against a run of `pod upgrade` and is unnecessary for `pod install` then. Unless the project isn't checking in pods and doesn't check in the `podfile.lock` file.

**Comment by bobgilmore** [comment]
Taking @Benuuu 's comment further, I'd like to see a succinct list of recommendations on top, along the lines of

* Check in your `Podfile`. 
* Check in your `Podfile.lock`. 
* Check in your `Pods/` directory.

The rationale and discussion is useful, but I think it would be best to cut to the chase and state our SOP up front.

**Comment by petey3** [comment]
Do we think its important to outline our expectations on what to do if you need to modify a pod? In my experience, that has meant forking a pod repo and then pointing to the intrepid fork in the `Podfile`

**Comment by alexpersian** [comment]
@petey3 wondering it that would be more suited to the [Cocoapods Processes](https://github.com/IntrepidPursuits/sherpa/blob/master/ios/cocoapods_processes.md) document? Otherwise it could definitely be added here to help out.

**Comment by petey3** [comment]
@alexpersian hmm yeah idk. IMO it might be better here if this is the "Best Practices" doc because the situation I am addressing is modifying someone else's pod for your project, as opposed to making your own pod which the [Cocoapods Processes](https://github.com/IntrepidPursuits/sherpa/blob/master/ios/cocoapods_processes.md) seems targeted towards

**Comment by alexpersian** [comment]
> Do we think its important to outline our expectations on what to do if you need to modify a pod? In my experience, that has meant forking a pod repo and then pointing to the intrepid fork in the `Podfile`

@petey3 would you be up for adding in that information to this doc once it's merged in? Sounds like you have more experience with what those best practices would be than I do.

**Review by paulrolfe — APPROVED** [review]
I think @alexpersian satisfied all the comments. I'm rebasing and merging this.

**Pull request merged**